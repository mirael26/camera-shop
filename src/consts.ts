export const AppUrl = {
  Main: '/',
  Catalog: '/catalog',
  Page: '/page_',
  Product: '/product',
  Cart: '/cart',
  NotFound: '/404',
  UnknownError: '/unknown_error',
  ServerUnavailable: '/server_unavailable',
} as const;

export const ApiUrl = {
  Promo: '/promo',
  Products: '/cameras',
  Reviews: '/reviews',
  Similar: '/similar',
  Promocode: '/coupons',
  Order: '/orders',
} as const;

export const Modal = {
  AddToCart: 'addToCartModal',
  DeleteFromCart: 'deleteFromCartModal',
  Review: 'reviewModal',
  OrderSuccess: 'orderSuccessModal',
} as const;

export const Tab = {
  Features: 'features',
  Description: 'description',
} as const;

export const Month = {
  0: 'января',
  1: 'февраля',
  2: 'марта',
  3: 'апреля',
  4: 'мая',
  5: 'июня',
  6: 'июля',
  7: 'августа',
  8: 'сентября',
  9: 'октября',
  10: 'ноября',
  11: 'декабря',
} as const;

export const Param = {
  Sort: 'sort',
  Order: 'order',
  Page: 'page',
  Tab: 'tab',
  PriceMin: 'price_min',
  PriceMax: 'price_max',
  Category: 'category',
  Level: 'level',
  Type: 'type',
} as const;

export const SortType = {
  Price: 'price',
  Rating: 'rating',
} as const;

export const SortOrder = {
  Asc: 'asc',
  Desc: 'desc',
} as const;

export const Category = {
  Camera: 'Фотоаппарат',
  VideoCamera: 'Видеокамера',
} as const;

export const Level = {
  Beginner: 'Нулевой',
  Amateur: 'Любительский',
  Professional: 'Профессиональный',
} as const;

export const CameraType = {
  Digital: 'Цифровая',
  Film: 'Плёночная',
  Instant: 'Моментальная',
  Collectible: 'Коллекционная',
} as const;
