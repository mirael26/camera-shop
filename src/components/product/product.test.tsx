import { fireEvent, getByText, screen } from '@testing-library/react';
import axios from 'axios';
import { useParams } from 'react-router-dom';
import { AppUrl, Tab } from '../../consts';
import { renderTestApp } from '../../test/helpers/render-test-app';
import { renderWithReduxAndRouter } from '../../test/helpers/render-with-redux-and-router';
import { productMock } from '../../test/mocks';
import Product from './product';

jest.mock('axios');
jest.mock('react-router-dom', () => {
  return {
    ...jest.requireActual('react-router-dom'),
    useParams: jest.fn(),
  }
});

describe('Product', () => {
  beforeEach(() => {
    jest.mocked(useParams).mockReturnValue({ id: '4' });
    jest.mocked(axios).get.mockResolvedValue(productMock);
  });

  test('Render correctly', () => {
    const product = renderWithReduxAndRouter(<Product/>, { initialState: { 
      data: { currentProduct: productMock }
    }});
    expect(product).toMatchSnapshot();
  });

  test('Features tab works correctly', () => {
    jest.mocked(useParams).mockReturnValueOnce({ id: '4', tab: Tab.Features });

    renderWithReduxAndRouter(<Product/>, { initialState: { 
      data: { currentProduct: productMock }
    }});
    
    expect(screen.getByText<HTMLLinkElement>(/Характеристики/i)).toHaveClass('is-active');
    expect(screen.getByText<HTMLLinkElement>(/Описание/i)).not.toHaveClass('is-active');
    expect(screen.getByTestId('features-content')).toHaveClass('is-active');
    expect(screen.getByTestId('description-content')).not.toHaveClass('is-active');
  });

  test('Description tab works correctly', () => {
    jest.mocked(useParams).mockReturnValue({ id: '4', tab: Tab.Description });
    renderWithReduxAndRouter(<Product/>, { initialState: { 
      data: { currentProduct: productMock }
    }});
    expect(screen.getByText<HTMLLinkElement>(/Характеристики/i)).not.toHaveClass('is-active');
    expect(screen.getByText<HTMLLinkElement>(/Описание/i)).toHaveClass('is-active');
    expect(screen.getByTestId('features-content')).not.toHaveClass('is-active');
    expect(screen.getByTestId('description-content')).toHaveClass('is-active');
  });

  test('AddToCart button invoke modal', () => {
    jest.mocked(useParams).mockReturnValue({ id: '4', tab: Tab.Description });
    renderTestApp(null, { route: `${AppUrl.Catalog}${AppUrl.Product}/4`,
      initialState: { 
        data: { currentProduct: productMock }
      }
    });
    const product = screen.getByTestId('product');
    const toCartButton = getByText(product, /Добавить в корзину/i);
    fireEvent.click(toCartButton);
    expect(screen.getByTestId('add-to-cart-modal')).toBeInTheDocument();
  });
});