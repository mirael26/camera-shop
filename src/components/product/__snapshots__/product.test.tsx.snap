// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="page-content__section"
      >
        <section
          class="product"
          data-testid="product"
        >
          <div
            class="container"
          >
            <div
              class="product__img"
            >
              <picture>
                <source
                  srcset="img/content/orlenok.webp, img/content/orlenok@2x.webp 2x"
                  type="image/webp"
                />
                <img
                  alt="img/content/orlenok.webp img/content/orlenok@2x.webp"
                  height="480"
                  src="img/content/orlenok.jpg"
                  srcset="img/content/orlenok@2x.jpg 2x"
                  width="560"
                />
              </picture>
            </div>
            <div
              class="product__content"
            >
              <h1
                class="title title--h3"
              >
                Орлёнок
              </h1>
              <div
                class="rate product__rate"
              >
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    xlink:href="#icon-full-star"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    xlink:href="#icon-full-star"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    xlink:href="#icon-full-star"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    xlink:href="#icon-star"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  height="16"
                  width="17"
                >
                  <use
                    xlink:href="#icon-star"
                  />
                </svg>
                <p
                  class="visually-hidden"
                >
                  Рейтинг: 
                  3
                </p>
                <p
                  class="rate__count"
                >
                  <span
                    class="visually-hidden"
                  >
                    Всего оценок:
                  </span>
                  17
                </p>
              </div>
              <p
                class="product__price"
              >
                <span
                  class="visually-hidden"
                >
                  Цена:
                </span>
                19 970
                 ₽
              </p>
              <button
                class="btn btn--purple"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  height="16"
                  width="24"
                >
                  <use
                    xlink:href="#icon-add-basket"
                  />
                </svg>
                Добавить в корзину
              </button>
              <div
                class="tabs product__tabs"
              >
                <div
                  class="tabs__controls product__tabs-controls"
                >
                  <a
                    class="tabs__control"
                    href="/catalog/product/4/features"
                  >
                    Характеристики
                  </a>
                  <a
                    class="tabs__control is-active"
                    href="/catalog/product/4/description"
                  >
                    Описание
                  </a>
                </div>
                <div
                  class="tabs__content"
                >
                  <div
                    class="tabs__element"
                    data-testid="features-content"
                  >
                    <ul
                      class="product__tabs-list"
                    >
                      <li
                        class="item-list"
                      >
                        <span
                          class="item-list__title"
                        >
                          Артикул:
                        </span>
                        <p
                          class="item-list__text"
                        >
                           
                          O78DFGSD83
                        </p>
                      </li>
                      <li
                        class="item-list"
                      >
                        <span
                          class="item-list__title"
                        >
                          Категория:
                        </span>
                        <p
                          class="item-list__text"
                        >
                          Фотоаппарат
                        </p>
                      </li>
                      <li
                        class="item-list"
                      >
                        <span
                          class="item-list__title"
                        >
                          Тип камеры:
                        </span>
                        <p
                          class="item-list__text"
                        >
                          Плёночная
                        </p>
                      </li>
                      <li
                        class="item-list"
                      >
                        <span
                          class="item-list__title"
                        >
                          Уровень:
                        </span>
                        <p
                          class="item-list__text"
                        >
                          Любительский
                        </p>
                      </li>
                    </ul>
                  </div>
                  <div
                    class="tabs__element is-active"
                    data-testid="description-content"
                  >
                    <div
                      class="product__tabs-text"
                    >
                      Плёночная перезаряжаемая камера нового покаления уже укомплектована плёнкой и оснащена встроенной вспышкой. Легко помещается в руке и обладет интересным дизайном.
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="page-content__section"
    >
      <section
        class="product"
        data-testid="product"
      >
        <div
          class="container"
        >
          <div
            class="product__img"
          >
            <picture>
              <source
                srcset="img/content/orlenok.webp, img/content/orlenok@2x.webp 2x"
                type="image/webp"
              />
              <img
                alt="img/content/orlenok.webp img/content/orlenok@2x.webp"
                height="480"
                src="img/content/orlenok.jpg"
                srcset="img/content/orlenok@2x.jpg 2x"
                width="560"
              />
            </picture>
          </div>
          <div
            class="product__content"
          >
            <h1
              class="title title--h3"
            >
              Орлёнок
            </h1>
            <div
              class="rate product__rate"
            >
              <svg
                aria-hidden="true"
                height="16"
                width="17"
              >
                <use
                  xlink:href="#icon-full-star"
                />
              </svg>
              <svg
                aria-hidden="true"
                height="16"
                width="17"
              >
                <use
                  xlink:href="#icon-full-star"
                />
              </svg>
              <svg
                aria-hidden="true"
                height="16"
                width="17"
              >
                <use
                  xlink:href="#icon-full-star"
                />
              </svg>
              <svg
                aria-hidden="true"
                height="16"
                width="17"
              >
                <use
                  xlink:href="#icon-star"
                />
              </svg>
              <svg
                aria-hidden="true"
                height="16"
                width="17"
              >
                <use
                  xlink:href="#icon-star"
                />
              </svg>
              <p
                class="visually-hidden"
              >
                Рейтинг: 
                3
              </p>
              <p
                class="rate__count"
              >
                <span
                  class="visually-hidden"
                >
                  Всего оценок:
                </span>
                17
              </p>
            </div>
            <p
              class="product__price"
            >
              <span
                class="visually-hidden"
              >
                Цена:
              </span>
              19 970
               ₽
            </p>
            <button
              class="btn btn--purple"
              type="button"
            >
              <svg
                aria-hidden="true"
                height="16"
                width="24"
              >
                <use
                  xlink:href="#icon-add-basket"
                />
              </svg>
              Добавить в корзину
            </button>
            <div
              class="tabs product__tabs"
            >
              <div
                class="tabs__controls product__tabs-controls"
              >
                <a
                  class="tabs__control"
                  href="/catalog/product/4/features"
                >
                  Характеристики
                </a>
                <a
                  class="tabs__control is-active"
                  href="/catalog/product/4/description"
                >
                  Описание
                </a>
              </div>
              <div
                class="tabs__content"
              >
                <div
                  class="tabs__element"
                  data-testid="features-content"
                >
                  <ul
                    class="product__tabs-list"
                  >
                    <li
                      class="item-list"
                    >
                      <span
                        class="item-list__title"
                      >
                        Артикул:
                      </span>
                      <p
                        class="item-list__text"
                      >
                         
                        O78DFGSD83
                      </p>
                    </li>
                    <li
                      class="item-list"
                    >
                      <span
                        class="item-list__title"
                      >
                        Категория:
                      </span>
                      <p
                        class="item-list__text"
                      >
                        Фотоаппарат
                      </p>
                    </li>
                    <li
                      class="item-list"
                    >
                      <span
                        class="item-list__title"
                      >
                        Тип камеры:
                      </span>
                      <p
                        class="item-list__text"
                      >
                        Плёночная
                      </p>
                    </li>
                    <li
                      class="item-list"
                    >
                      <span
                        class="item-list__title"
                      >
                        Уровень:
                      </span>
                      <p
                        class="item-list__text"
                      >
                        Любительский
                      </p>
                    </li>
                  </ul>
                </div>
                <div
                  class="tabs__element is-active"
                  data-testid="description-content"
                >
                  <div
                    class="product__tabs-text"
                  >
                    Плёночная перезаряжаемая камера нового покаления уже укомплектована плёнкой и оснащена встроенной вспышкой. Легко помещается в руке и обладет интересным дизайном.
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
